modules/videoio/src/cap_msmf.cpp

```diff
@@ -4023,261 +3976,147 @@ const GUID CvVideoWriter_MSMF::FourCC2GUID(int fourcc)
+                SUCCEEDED(MFSetAttributeRatio(mediaTypeOut.Get(), MF_MT_FRAME_RATE, (UINT32)fps, 1)) &&
+                SUCCEEDED(MFSetAttributeRatio(mediaTypeOut.Get(), MF_MT_PIXEL_ASPECT_RATIO, 1, 1)) &&
+                // Set the input media type.
+                SUCCEEDED(MFCreateMediaType(&mediaTypeIn)) &&
+                SUCCEEDED(mediaTypeIn->SetGUID(MF_MT_MAJOR_TYPE, MFMediaType_Video)) &&
+                SUCCEEDED(mediaTypeIn->SetGUID(MF_MT_SUBTYPE, inputFormat)) &&
+                SUCCEEDED(mediaTypeIn->SetUINT32(MF_MT_INTERLACE_MODE, MFVideoInterlace_Progressive)) &&
+                SUCCEEDED(mediaTypeIn->SetUINT32(MF_MT_DEFAULT_STRIDE, 4 * videoWidth)) && //Assume BGR32 input
+                SUCCEEDED(MFSetAttributeSize(mediaTypeIn.Get(), MF_MT_FRAME_SIZE, videoWidth, videoHeight)) &&
+                SUCCEEDED(MFSetAttributeRatio(mediaTypeIn.Get(), MF_MT_FRAME_RATE, (UINT32)fps, 1)) &&
+                SUCCEEDED(MFSetAttributeRatio(mediaTypeIn.Get(), MF_MT_PIXEL_ASPECT_RATIO, 1, 1)) &&
+                // Set sink writer parameters
+                SUCCEEDED(MFCreateAttributes(&spAttr, 10)) &&
+                SUCCEEDED(spAttr->SetUINT32(MF_READWRITE_ENABLE_HARDWARE_TRANSFORMS, true)) &&
+                SUCCEEDED(spAttr->SetUINT32(MF_SINK_WRITER_DISABLE_THROTTLING, true))
+                )
             {
-                initiated = true;
-                rtStart = 0;
-                MFFrameRateToAverageTimePerFrame((UINT32)fps, 1, &rtDuration);
+                // Create the sink writer
+                wchar_t* unicodeFileName = new wchar_t[strlen(filename) + 1];
+                MultiByteToWideChar(CP_ACP, 0, filename, -1, unicodeFileName, (int)strlen(filename) + 1);
+                HRESULT hr = MFCreateSinkWriterFromURL(unicodeFileName, NULL, spAttr.Get(), &sinkWriter);
+                delete[] unicodeFileName;
+                if (SUCCEEDED(hr))
+                {
+                    // Configure the sink writer and tell it start to start accepting data
+                    if (SUCCEEDED(sinkWriter->AddStream(mediaTypeOut.Get(), &streamIndex)) &&
+                        SUCCEEDED(sinkWriter->SetInputMediaType(streamIndex, mediaTypeIn.Get(), NULL)) &&
+                        SUCCEEDED(sinkWriter->BeginWriting()))
+                    {
+                        initiated = true;
+                        rtStart = 0;
+                        MFFrameRateToAverageTimePerFrame((UINT32)fps, 1, &rtDuration);
+                        return true;
+                    }
+                }
             }
+            MFShutdown();
         }
+        CoUninitialize();
     }
-
-    return SUCCEEDED(hr);
+    return false;
 }
 
 void CvVideoWriter_MSMF::close()
 {
-    if (!initiated)
+    if (initiated)
     {
-        return;
+        initiated = false;
+        sinkWriter->Finalize();
+        SafeRelease(&sinkWriter);
+        MFShutdown();
+        CoUninitialize();
```

Comment:
> MFShutdown();
> CoUninitialize();

We should not call global cleanup/unload functions.

The same note is for other places.